pkg/webui/console/containers/device-onboarding-form/device-qr-scan-form-section/index.js

```diff
@@ -0,0 +1,191 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React, { useCallback, useState } from 'react'
+import { defineMessages } from 'react-intl'
+import { useSelector, useDispatch } from 'react-redux'
+
+import QRModalButton from '@ttn-lw/components/qr-model-button'
+import { useFormContext } from '@ttn-lw/components/form'
+import Link from '@ttn-lw/components/link'
+import Icon from '@ttn-lw/components/icon'
+import ModalButton from '@ttn-lw/components/button/modal-button'
+
+import Message from '@ttn-lw/lib/components/message'
+
+import sharedMessages from '@ttn-lw/lib/shared-messages'
+
+import { readDeviceQr } from '@console/lib/device-claiming-parse-qr'
+
+import { listBrands } from '@console/store/actions/device-repository'
+
+import { selectDeviceBrands } from '@console/store/selectors/device-repository'
+import { selectSelectedApplicationId } from '@console/store/selectors/applications'
+
+const hexToDecimal = hex => parseInt(hex, 16)
+
+const m = defineMessages({
+  hasEndDeviceQR: 'Does your end device have a QR code? Scan it to speed registration.',
+  learnMore: 'Learn more',
+  scanEndDevice: 'Scan end device code',
+  deviceInfo: 'Device information',
+  resetQRCodeData: 'Reset QR code data',
+  resetConfirm:
+    'Are you sure you want to discard QR code data? The scanned device will not be registered and the form will be reset',
+  scanSuccess: 'QR code scanned successfully',
+})
+
+const qrDataInitialValues = {
+  valid: false,
+  approved: false,
+  data: undefined,
+  device: undefined,
+}
+
+const DeviceQRScanFormSection = () => {
+  const dispatch = useDispatch()
+  const brands = useSelector(selectDeviceBrands)
+  const appId = useSelector(selectSelectedApplicationId)
+  const [qrData, setQrData] = useState(qrDataInitialValues)
+
+  React.useEffect(() => {
+    dispatch(listBrands(appId, {}, ['name', 'lora_alliance_vendor_id']))
+  }, [appId, dispatch])
+
+  const { setFieldValue, values, resetForm } = useFormContext()
+
+  const handleReset = useCallback(() => {
+    resetForm()
+    setQrData(qrDataInitialValues)
+  }, [resetForm])
+
+  const getBrand = useCallback(
+    vendorId => {
+      const brand = brands.filter(
+        brand => brand?.lora_alliance_vendor_id === hexToDecimal(vendorId),
+      )
+
+      return brand[0]
+    },
+    [brands],
+  )
+
+  const handleQRCodeApprove = useCallback(() => {
+    const { device } = qrData
+
+    if (device.profileID.vendorID && device.profileID.vendorID !== '0000') {
+      const brand = getBrand(device.profileID.vendorID)
+      setFieldValue('version_ids', {
+        ...values.version_ids,
+        brand_id: brand.brand_id,
+      })
+    }
+
+    setFieldValue('ids', {
+      ...values.ids,
+      join_eui: device.joinEUI,
+      dev_eui: device.devEUI,
+    })
```

Comment:
```suggestion
    setValues(values => ({
      ...values,
      ids: {
        ..values.ids,
        join_eui: device.joinEUI,
        dev_eui: device.devEUI,
      },
    })
```

If you use values from the context, they are always one render-cycle behind and can hence be outdated by the time you try to update them. Thus it's better to always use the factory function.